authority/provisioner/webhook.go

```diff
@@ -0,0 +1,204 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"text/template"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/templates"
+	"go.step.sm/crypto/sshutil"
+	"golang.org/x/crypto/ssh"
+)
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	SigningSecret        string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+type sshRequest struct {
+	Key        []byte
+	Type       string
+	KeyID      string
+	Principals []string
+}
+
+// ssh.Certificate but the Key and SignatureKey are marshaled
+type sshCert struct {
+	Nonce    []byte
+	Key      []byte
+	Serial   uint64
+	CertType uint32
+	//nolint:revive // field name matches ssh.Certificate
+	KeyId           string
+	ValidPrincipals []string
+	ValidAfter      uint64
+	ValidBefore     uint64
+	Permissions     ssh.Permissions
+	Reserved        []byte
+	SignatureKey    []byte
+	Signature       *ssh.Signature
+}
+
+type webhookRequestBody struct {
+	Timestamp string `json:"timestamp"`
+	//nolint:revive // acronyms can use all caps
+	X509_CSR []byte `json:"csr,omitempty"`
+	//nolint:revive // acronyms can use all caps
+	SSH_CR         *sshRequest       `json:"ssh_cr,omitempty"`
+	Certificate    *x509.Certificate `json:"certificate,omitempty"`
```

Comment:
I can add something like that in crypto, another option is using the template representation.

Reply:
I think the template representation is a good option since users expect that to be the json format of a certificate. There's already a function to [get the x509 cert from the template](https://github.com/smallstep/crypto/blob/master/x509util/certificate.go#L88). I'll add a function in crypto to get the template from the x509.

Reply:
2ee14b7d